File: Problems/P0054/experiments/g4.12-exact-fifty-four/middle-seven-nine-six-row.cpp
#include <algorithm>
#include <array>
#include <bit>
#include <fstream>
#include <iostream>
#include <map>
#include <set>
#include <sstream>
#include <string>
#include <vector>

constexpr int FULL = 63;

using Profile = std::vector<int>;

int transform(int support, const std::array<int, 3> &permutation)
{
	int image = 0;
	for (int row = 0; row < 3; ++row) {
		if (support & (1 << row)) image |= 1 << permutation[row];
	}
	return image;
}

Profile canonical_profile(const std::array<int, 7> &multiplicity)
{
	Profile best;
	bool has_best = false;
	std::array<int, 3> permutation{0, 1, 2};
	do {
		Profile current;
		for (int support = 1; support < 7; ++support) {
			for (int copy = 0; copy < multiplicity[support]; ++copy) {
				current.push_back(transform(support, permutation));
			}
		}
		std::sort(current.begin(), current.end());
		if (!has_best || current < best) {
			best = current;
			has_best = true;
		}
	} while (std::next_permutation(permutation.begin(), permutation.end()));
	return best;
}

std::string profile_key(const Profile &profile)
{
	std::ostringstream output;
	for (int support = 1; support < 7; ++support) {
		int count = std::count(profile.begin(), profile.end(), support);
		if (count) output << support << ':' << count << ',';
	}
	return output.str();
}

int pair_blockers(const Profile &supports)
{
	int count = 0;
	for (int left = 0; left < static_cast<int>(supports.size()); ++left) {
		for (int right = left + 1; right < static_cast<int>(supports.size()); ++right) {
			if ((supports[left] | supports[right]) == FULL) ++count;
		}
	}
	return count;
}

void enumerate_matchings(
	const Profile &left,
	const Profile &right,
	int position,
	std::vector<bool> &used,
	Profile &combined,
	std::map<int, std::uint64_t> &histogram
)
{
	if (position == static_cast<int>(left.size())) {
		Profile supports{7, 56};
		supports.insert(supports.end(), combined.begin(), combined.end());
		for (int index = 0; index < static_cast<int>(right.size()); ++index) {
			if (!used[index]) supports.push_back(right[index] << 3);
		}
		++histogram[pair_blockers(supports)];
		return;
	}

	combined.push_back(left[position]);
	enumerate_matchings(left, right, position + 1, used, combined, histogram);
	combined.pop_back();

	for (int index = 0; index < static_cast<int>(right.size()); ++index) {
		if (used[index]) continue;
		if (std::popcount(static_cast<unsigned>(left[position]))
			+ std::popcount(static_cast<unsigned>(right[index])) > 4) {
			continue;
		}
		used[index] = true;
		combined.push_back(left[position] | (right[index] << 3));
		enumerate_matchings(left, right, position + 1, used, combined, histogram);
		combined.pop_back();
		used[index] = false;
	}
}

int main(int argc, char **argv)
{
	if (argc != 2) {
		std::cerr << "usage: middle-seven-nine-six-row OUTPUT\n";
		return 2;
	}

	std::set<Profile> profiles;
	std::uint64_t labelled_solutions = 0;
	for (int p = 0; p <= 5; ++p) {
		for (int q = 0; q <= 5; ++q) {
			for (int r = 0; r <= 5; ++r) {
				for (int x = 0; x <= 5; ++x) {
					for (int y = 0; y <= 5; ++y) {
						for (int z = 0; z <= 5; ++z) {
							std::array<int, 3> sizes{p + x + y, q + x + z, r + y + z};
							if (std::ranges::any_of(sizes, [](int size) {
								return size < 2 || size > 5;
							})) continue;
							if (!(p + y) || !(q + z) || !(p + x) || !(r + z)
								|| !(q + x) || !(r + y)) {
								continue;
							}
							int pairs = x * y + x * z + y * z + x * r + y * q + z * p;
							int triples = p * q * r;
							if (pairs + triples > 4) continue;
							++labelled_solutions;
							std::array<int, 7> multiplicity{};
							multiplicity[1] = p;
							multiplicity[2] = q;
							multiplicity[4] = r;
							multiplicity[3] = x;
							multiplicity[5] = y;
							multiplicity[6] = z;
							profiles.insert(canonical_profile(multiplicity));
						}
					}
				}
			}
		}
	}

	std::map<int, std::uint64_t> histogram;
	for (Profile left : profiles) {
		left.push_back(7);
		for (Profile right : profiles) {
			right.push_back(7);
			std::vector<bool> used(right.size());
			Profile combined;
			enumerate_matchings(left, right, 0, used, combined, histogram);
		}
	}

	std::ofstream output(argv[1]);
	if (!output) {
		std::cerr << "cannot open output\n";
		return 2;
	}
	output << "schema=p0054.g4.12.middle-seven-nine-six-row.v1\n";
	output << "marginal_labelled_solutions=" << labelled_solutions << '\n';
	output << "marginal_row_permutation_classes=" << profiles.size() << '\n';
	int class_index = 0;
	for (const Profile &profile : profiles) {
		output << "class_" << ++class_index << '=' << profile_key(profile) << '\n';
	}
	std::uint64_t couplings = 0;
	for (const auto &[pairs, count] : histogram) {
		output << "pair_blockers_" << pairs << '=' << count << '\n';
		couplings += count;
	}
	output << "labelled_couplings=" << couplings << '\n';
	output << "minimum_pair_blockers=" << histogram.begin()->first << '\n';
	std::cout << "output=" << argv[1] << '\n';
	return 0;
}

File: Problems/P0054/experiments/g4.12-exact-fifty-four/middle-seven-nine-six-row.py
#!/usr/bin/env python3
"""Independently verify the G4.12 six-row coupling enumeration."""

from __future__ import annotations

import argparse
import hashlib
import itertools
import json
from collections import Counter
from collections.abc import Iterator
from pathlib import Path

Profile = tuple[int, ...]


def sha256_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1 << 20), b""):
            digest.update(chunk)
    return digest.hexdigest()


def transform(support: int, permutation: tuple[int, ...]) -> int:
    return sum(1 << permutation[row] for row in range(3) if support & (1 << row))


def canonical_profile(multiplicity: dict[int, int]) -> Profile:
    return min(
        tuple(
            sorted(
                transform(support, permutation)
                for support, count in multiplicity.items()
                for _ in range(count)
            )
        )
        for permutation in itertools.permutations(range(3))
    )


def marginal_profiles() -> tuple[int, tuple[Profile, ...]]:
    labelled = 0
    classes: set[Profile] = set()
    for p, q, r, x, y, z in itertools.product(range(6), repeat=6):
        sizes = (p + x + y, q + x + z, r + y + z)
        if any(not 2 <= size <= 5 for size in sizes):
            continue
        if not (p + y and q + z and p + x and r + z and q + x and r + y):
            continue
        pairs = x * y + x * z + y * z + x * r + y * q + z * p
        triples = p * q * r
        if pairs + triples > 4:
            continue
        labelled += 1
        classes.add(canonical_profile({1: p, 2: q, 4: r, 3: x, 5: y, 6: z}))
    return labelled, tuple(sorted(classes))


def partial_couplings(
    left: Profile,
    right: Profile,
    position: int = 0,
    used: frozenset[int] = frozenset(),
) -> Iterator[Profile]:
    if position == len(left):
        yield tuple(right[index] << 3 for index in range(len(right)) if index not in used)
        return
    for suffix in partial_couplings(left, right, position + 1, used):
        yield (left[position],) + suffix
    for index, support in enumerate(right):
        if index in used or left[position].bit_count() + support.bit_count() > 4:
            continue
        for suffix in partial_couplings(left, right, position + 1, used | {index}):
            yield (left[position] | (support << 3),) + suffix


def pair_blockers(supports: Profile) -> int:
    return sum(left | right == 63 for left, right in itertools.combinations(supports, 2))


def parse_reference(path: Path) -> dict[str, str]:
    return dict(
        line.split("=", 1) for line in path.read_text(encoding="ascii").splitlines() if "=" in line
    )


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--reference", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()

    labelled, profiles = marginal_profiles()
    histogram: Counter[int] = Counter()
    for left in profiles:
        for right in profiles:
            for coupling in partial_couplings(left + (7,), right + (7,)):
                histogram[pair_blockers((7, 56) + coupling)] += 1

    expected = {
        "schema": "p0054.g4.12.middle-seven-nine-six-row.v1",
        "marginal_labelled_solutions": str(labelled),
        "marginal_row_permutation_classes": str(len(profiles)),
        **{
            f"class_{index}": ",".join(
                f"{support}:{profile.count(support)}" for support in sorted(set(profile))
            )
            + ","
            for index, profile in enumerate(profiles, 1)
        },
        **{f"pair_blockers_{pairs}": str(count) for pairs, count in sorted(histogram.items())},
        "labelled_couplings": str(sum(histogram.values())),
        "minimum_pair_blockers": str(min(histogram)),
    }
    reference = parse_reference(args.reference)
    if reference != expected:
        raise AssertionError("independent six-row enumeration disagrees with reference")

    report = {
        "schema": "p0054.g4.12.middle-seven-nine-six-row-verification.v1",
        "epistemic_label": "COMPUTED",
        "reference_path": str(args.reference),
        "reference_sha256": sha256_file(args.reference),
        "python_source_sha256": sha256_file(Path(__file__)),
        "marginal_labelled_solutions": labelled,
        "marginal_row_permutation_classes": len(profiles),
        "labelled_couplings": sum(histogram.values()),
        "pair_blocker_histogram": {str(pairs): count for pairs, count in sorted(histogram.items())},
        "minimum_pair_blockers": min(histogram),
        "reference_match": True,
    }
    args.output.write_text(json.dumps(report, indent=2, sort_keys=True) + "\n", encoding="ascii")
    print(json.dumps(report, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()
